src/borrow_tracker/tree_borrows/mod.rs

```diff
@@ -342,35 +342,80 @@ trait EvalContextPrivExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         assert!(new_perm.freeze_access);
 
         let protected = new_perm.protector.is_some();
-        this.visit_freeze_sensitive(place, ptr_size, |range, frozen| {
-            has_unsafe_cell = has_unsafe_cell || !frozen;
-
-            // We are only ever `Frozen` inside the frozen bits.
-            let (perm, access) = if frozen {
+        let precise_interior_mut = this
+            .machine
+            .borrow_tracker
+            .as_ref()
+            .unwrap()
+            .borrow()
```

Comment:
```suggestion
            .get_mut()
```
Let's avoid the run-time checks of `borrow()`, we don't need them here.

Reply:
For this to work, I had to move down

```rust
        let alloc_extra = this.get_alloc_extra(alloc_id)?;
        let mut tree_borrows = alloc_extra.borrow_tracker_tb().borrow_mut();
```

from line 315.